src/middleware.ts

```diff
@@ -259,43 +296,59 @@ export const persist = <S extends State>(
   }
 
   // rehydrate initial state with existing stored state
-  ;(async () => {
-    const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
-
-    try {
-      const storageValue = await storage.getItem(name)
 
+  // a workaround to solve the issue of not storing rehydrated state in sync storage
+  // the set(state) value would be later overridden with initial state by create()
+  // to avoid this, we merge the state from localStorage into the initial state.
+  // TODO: find a better solution for this
+  let stateFromStorage: any
```

Comment:
Okay, I understand why this is required. I think it's the right solution.
Let's rename it like:
```suggestion
  let stateFromStorageInSync: S | undefined
```

The type `S` should probably be `Partial<S>` not only here, but all over the place. Maybe we should tackle it in a separate PR.